python/fnllm/fnllm/openai/services/openai_retryable_error_handler.py

```diff
@@ -21,25 +30,50 @@
 ]
 
 
+class OpenAIBackoffLimiter(Limiter):
+    """
+    OpenAI Backoff Limiter.
+
+    This limiter is used in tandem with the OpenAI Reetryable Error Handler to ensure that when
+    429 errors are encountered, the running loop will pause for the amount of time specified in the
+    retry-after header before allowing requests to clear the limiters.
+
+    Any in-flight requests will be allowed to complete, but no new requests will be allowed to start
+    until the delay has passed.
+    """
+
+    def __init__(self) -> None:
+        """Create a new OpenAIBackoffLimiter."""
+        self._delay_event = asyncio.Event()  # Event to control access
+        self._delay_event.set()  # Initially allow acquire() calls
+
+    async def acquire(self, manifest: Manifest) -> None:
+        """Acquire a pass through the limiter."""
+        await self._delay_event.wait()  # Wait until the event is set
+
+    async def release(self, manifest: Manifest) -> None:
+        """Release a pass through the limiter."""
+        # No-op in this implementation, but can be extended if needed.
+
+    async def sleep_for(self, time: int) -> None:
+        """Sleep for the given amount of time."""
+        self._delay_event.clear()  # Block acquire() calls
+        try:
+            await asyncio.sleep(time)
+        finally:
+            self._delay_event.set()
+
+
 class OpenAIRetryableErrorHandler:
     """A base class to rate limit the LLM."""
 
+    def __init__(self, limiter: OpenAIBackoffLimiter) -> None:
+        """Create a new OpenAIRetryableErrorHandler."""
+        self._limiter = limiter
+
     async def __call__(self, error: BaseException) -> None:
         """Handle the rate limit error."""
-        sleep_recommendation = self._extract_sleep_recommendation(error)
-        if sleep_recommendation > 0:
-            await asyncio.sleep(sleep_recommendation)
-
-    def _extract_sleep_recommendation(self, error: BaseException) -> float:
-        """Extract the sleep time value from a RateLimitError. This is usually only available in Azure."""
-        please_retry_after_msg: Final = "Rate limit is exceeded. Try again in "
-        error_str = str(error)
-
-        if (
-            not isinstance(error, RateLimitError)
-            or please_retry_after_msg not in error_str
-        ):
-            return 0
-
-        # could be second or seconds
-        return int(error_str.split(please_retry_after_msg)[1].split(" second")[0])
+        if isinstance(error, APIStatusError):
```

Comment:
you could also use a match/case here, I like the match/case syntax better than the if isinstance() but it's a personal opinion :) 

```py
match error:
  case APIStatusError():
    retry_after = error.response.headers.get("retry-after", None)
    if retry_after is not None:
        await self._limiter.sleep_for(int(retry_after))
  case _:
    pass # or log something in case we want to track errors that are not handled 

```